sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java

```diff
@@ -112,73 +109,60 @@ private FlatMapElements(
-
-              @Override
-              public TypeDescriptor<OutputT> getOutputTypeDescriptor() {
-                @SuppressWarnings({"rawtypes", "unchecked"}) // safe by static typing
-                TypeDescriptor<Iterable<?>> iterableType =
-                    (TypeDescriptor) fn.getOutputTypeDescriptor();
-
-                @SuppressWarnings("unchecked") // safe by correctness of getIterableElementType
-                TypeDescriptor<OutputT> outputType =
-                    (TypeDescriptor<OutputT>) getIterableElementType(iterableType);
-
-                return outputType;
-              }
-            }));
+    checkArgument(fn != null, ".via() is required");
+    PCollection<OutputT> res =
+        input.apply(
+            "Map",
+            ParDo.of(
+                    new DoFn<InputT, OutputT>() {
+                      @ProcessElement
+                      public void processElement(ProcessContext c) throws Exception {
+                        Iterable<OutputT> res =
+                            fn.getClosure().apply(c.element(), Fn.Context.wrapProcessContext(c));
+                        for (OutputT output : res) {
+                          c.output(output);
+                        }
+                      }
+
+                      @Override
+                      public void populateDisplayData(DisplayData.Builder builder) {
+                        builder.delegate(FlatMapElements.this);
+                      }
+                    })
+                .withSideInputs(fn.getRequirements().getSideInputs()));
+    res.setTypeDescriptor(outputType);
+    try {
+      Coder<OutputT> outputCoder =
+          input
+              .getPipeline()
+              .getCoderRegistry()
+              .getCoder(outputType, inputType, (Coder<InputT>) input.getCoder());
+      res.setCoder(outputCoder);
+    } catch (CannotProvideCoderException e) {
+      // Ignore
+    }
+    return res;
   }
 
   @Override
   public void populateDisplayData(DisplayData.Builder builder) {
     super.populateDisplayData(builder);
-    builder
-        .include("flatMapFn", fn)
-        .add(fnClassDisplayData);
-  }
-
-  /**
-   * Does a best-effort job of getting the best {@link TypeDescriptor} for the type of the
-   * elements contained in the iterable described by the given {@link TypeDescriptor}.
-   */
-  private static TypeDescriptor<?> getIterableElementType(
-      TypeDescriptor<Iterable<?>> iterableTypeDescriptor) {
-
-    // If a rawtype was used, the type token may be for Object, not a subtype of Iterable.
-    // In this case, we rely on static typing of the function elsewhere to ensure it is
-    // at least some kind of iterable, and grossly overapproximate the element type to be Object.
-    if (!iterableTypeDescriptor.isSubtypeOf(new TypeDescriptor<Iterable<?>>() {})) {
-      return new TypeDescriptor<Object>() {};
+    Fn<InputT, ? extends Iterable<OutputT>> closure = fn.getClosure();
+    if (closure instanceof HasDisplayData) {
```

Comment:
Should we add the class if the closure doesn't implement HasDisplayData?
(Ditto for MapElements)